File: 2025/04 April/12 Flood fill Algorithm/main.cpp
// Link: https://www.geeksforgeeks.org/problems/flood-fill-algorithm1856/1

#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution
{
public:
    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc, int newColor)
    {

        if (image[sr][sc] == newColor)
        {
            return image;
        }

        vector<pair<int, int>> directions = {
            {1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        queue<pair<int, int>> q;
        int oldColor = image[sr][sc];
        q.push({sr, sc});

        image[sr][sc] = newColor;

        while (!q.empty())
        {
            pair<int, int> front = q.front();
            int x = front.first, y = front.second;
            q.pop();

            for (const pair<int, int> &direction : directions)
            {
                int nx = x + direction.first;
                int ny = y + direction.second;

                if (nx >= 0 && nx < image.size() &&
                    ny >= 0 && ny < image[0].size() &&
                    image[nx][ny] == oldColor)
                {

                    image[nx][ny] = newColor;
                    q.push({nx, ny});
                }
            }
        }

        return image;
    }
};

int main()
{
    Solution sol;
    vector<vector<int>> image = {{1, 1, 1, 0}, {0, 1, 1, 1}, {1, 0, 1, 1}};
    int sr = 1;
    int sc = 2;
    int newColor = 2;
    vector<vector<int>> result = sol.floodFill(image, sr, sc, newColor);
    for (const auto &row : result)
    {
        for (const auto &col : row)
        {
            cout << col << " ";
        }
        cout << endl;
    }
    cout << endl;

    image = {{1, 1, 1}, {1, 1, 0}, {1, 0, 1}};
    sr = 1;
    sc = 1;
    newColor = 2;
    result = sol.floodFill(image, sr, sc, newColor);
    for (const auto &row : result)
    {
        for (const auto &col : row)
        {
            cout << col << " ";
        }
        cout << endl;
    }
    cout << endl;

    image = {{0, 1, 0}, {0, 1, 0}};
    sr = 0;
    sc = 1;
    newColor = 0;
    result = sol.floodFill(image, sr, sc, newColor);
    for (const auto &row : result)
    {
        for (const auto &col : row)
        {
            cout << col << " ";
        }
        cout << endl;
    }
    cout << endl;

    return 0;
}

File: 2025/04 April/12 Flood fill Algorithm/main.py
# Link: https://www.geeksforgeeks.org/problems/flood-fill-algorithm1856/1


class Solution:
    def bfs(self, sr, sc, image, ans, newcolor, initcolor):
        q = [[sr, sc]]
        ans[sr][sc] = newcolor
        row = len(image)
        col = len(image[0])
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        while q:
            i, j = q.pop(0)
            for dr, dc in directions:
                nrow = i + dr
                ncol = j + dc
                if (
                    0 <= nrow < row
                    and 0 <= ncol < col
                    and image[nrow][ncol] == initcolor
                    and ans[nrow][ncol] != newcolor
                ):
                    q.append([nrow, ncol])
                    ans[nrow][ncol] = newcolor

    def floodFill(self, image, sr, sc, newcolor):
        initcolor = image[sr][sc]
        if initcolor == newcolor:
            return image
        ans = [row[:] for row in image]
        self.bfs(sr, sc, image, ans, newcolor, initcolor)
        return ans


if __name__ == "__main__":
    sol = Solution()
    image = [[1, 1, 1, 0], [0, 1, 1, 1], [1, 0, 1, 1]]
    sr = 1
    sc = 2
    newColor = 2
    result = sol.floodFill(image, sr, sc, newColor)
    for row in result:
        print(row)
    print()

    image = [[1, 1, 1], [1, 1, 0], [1, 0, 1]]
    sr = 1
    sc = 1
    newColor = 2
    result = sol.floodFill(image, sr, sc, newColor)
    for row in result:
        print(row)
    print()

    image = [[0, 1, 0], [0, 1, 0]]
    sr = 0
    sc = 1
    newColor = 0
    result = sol.floodFill(image, sr, sc, newColor)
    for row in result:
        print(row)
    print()
